From 22e205342e6ee049b162761dc43bbd0576d87d4d Mon Sep 17 00:00:00 2001
From: Pardha Kesapragada <pkesapragada@habana.ai>
Date: Thu, 18 Mar 2021 11:49:31 +0200
Subject: [PATCH] [SW-30985] ppboot: add reset status decoding capability

In all platforms until Greco, PSOC_GLOBAL_CONF_RST_SRC
contained an index which will determine the cause of reset.
But for Gaudi2, more resets have been added, each with
a bit mask, with each reset denoting a specific case.
Hence add a capability to decode the reset causes per
platform basis. The platform specific resets
are mapped to a common structure which ppboot uses to
take appropriate actions

Change-Id: I1cc62fa2fb6d8c3c2937613ebfd92112280aff5b
Signed-off-by: Pardha Kesapragada <pkesapragada@habana.ai>
---

diff --git a/gaudi/common.c b/gaudi/common.c
index 957a49f..c348264 100644
--- a/gaudi/common.c
+++ b/gaudi/common.c
@@ -420,3 +420,33 @@
 	preboot_ack_razwi_irq();
 	preboot_security_dbi_lock();
 }
+
+u32 decode_reset_cause(void)
+{
+	u32 rst_src;
+
+	rst_src = regmap_readl(mmPSOC_GLOBAL_CONF_RST_SRC);
+
+	switch (rst_src) {
+	case GAUDI_COLD_RST:
+		rst_src = HL_COLD_RST;
+		break;
+	case GAUDI_PRSTN_RST:
+		rst_src = HL_PRSTN_RST;
+		break;
+	case GAUDI_MANUAL_RST:
+		rst_src = HL_MANUAL_RST;
+		break;
+	case GAUDI_WD_RST:
+		rst_src = HL_WD_RST;
+		break;
+	case GAUDI_SW_ALL_RST:
+		rst_src = HL_SW_ALL_RST;
+		break;
+	default:
+		rst_src = HL_UNKNOWN_RST;
+		break;
+	}
+
+	return rst_src;
+}
diff --git a/gaudi/common.h b/gaudi/common.h
index 4ab14a4..6d553ed 100644
--- a/gaudi/common.h
+++ b/gaudi/common.h
@@ -59,5 +59,13 @@
 int gaudi_hbm_init(void);
 u32 preboot_get_vendor_id_device_id(void);
 
+enum gaudi_rst_src {
+	GAUDI_COLD_RST,
+	GAUDI_MANUAL_RST,
+	GAUDI_PRSTN_RST = 2,
+	GAUDI_WD_RST = 4,
+	GAUDI_SW_ALL_RST = 8,
+};
+
 #endif /* __ASSEMBLY__ */
 #endif /* COMMON_GAUDI_H_ */
diff --git a/gaudi2_arc_gnu/common.c b/gaudi2_arc_gnu/common.c
index b5b8750..e1c3e85 100644
--- a/gaudi2_arc_gnu/common.c
+++ b/gaudi2_arc_gnu/common.c
@@ -454,3 +454,50 @@
 	*loadaddr = copy_preboot();
 	return true;
 }
+
+u32 decode_reset_cause(void)
+{
+	u32 pos, rst_src, reg;
+
+	reg = regmap_readl(mmPSOC_GLOBAL_CONF_RST_SRC);
+
+	if (!reg)
+		return HL_UNKNOWN_RST;
+
+	pos = (__builtin_ffs(reg) - 1);
+
+	switch (pos) {
+	case GAUDI2_COLD_RST:
+		rst_src = HL_COLD_RST;
+		break;
+	case GAUDI2_PRSTN_RST:
+		rst_src = HL_PRSTN_RST;
+		break;
+	case GAUDI2_MANUAL_RST:
+		rst_src = HL_MANUAL_RST;
+		break;
+	case GAUDI2_WD_RST:
+		rst_src = HL_WD_RST;
+		break;
+	case GAUDI2_SW_RST:
+		rst_src = HL_SW_ALL_RST;
+		break;
+	case GAUDI2_SOFT_RST:
+		rst_src = HL_SOFT_RST;
+		break;
+	case GAUDI2_FW_RST:
+		rst_src = HL_FW_RST;
+		break;
+	case GAUDI2_FLR_RST:
+		rst_src = HL_FLR_RST;
+		break;
+	case GAUDI2_ECC_DERR_RST:
+		rst_src = HL_ECC_DERR_RST;
+		break;
+	default:
+		rst_src = HL_UNKNOWN_RST;
+		break;
+	}
+
+	return rst_src;
+}
diff --git a/gaudi2_arc_gnu/common.h b/gaudi2_arc_gnu/common.h
index 9a9712e..9beb024 100644
--- a/gaudi2_arc_gnu/common.h
+++ b/gaudi2_arc_gnu/common.h
@@ -7,6 +7,7 @@
 #define COMMON_GAUDI2_ARC_GNU_H_
 
 #include <gaudi2_blocks_linux_driver.h>
+#include <psoc_global_conf_masks.h>
 #include <bitops.h>
 
 #define SRAM_SIZE			0x3000000	/* 48MB */
@@ -58,4 +59,19 @@
 #define ARC_RST_VALUE		0x110
 #define ARC_RST_VEC_ADDR	0x0
 
+#ifndef __ASSEMBLY__
+
+enum gaudi2_rst_src {
+	GAUDI2_COLD_RST = PSOC_GLOBAL_CONF_RST_SRC_COLD_RST_IND_SHIFT,
+	GAUDI2_MANUAL_RST = PSOC_GLOBAL_CONF_RST_SRC_MNL_RST_IND_SHIFT,
+	GAUDI2_PRSTN_RST = PSOC_GLOBAL_CONF_RST_SRC_PRSTN_RST_IND_SHIFT,
+	GAUDI2_SOFT_RST = PSOC_GLOBAL_CONF_RST_SRC_SOFT_RST_IND_SHIFT,
+	GAUDI2_WD_RST = PSOC_GLOBAL_CONF_RST_SRC_WD_RST_IND_SHIFT,
+	GAUDI2_FW_RST = PSOC_GLOBAL_CONF_RST_SRC_FW_RST_IND_SHIFT,
+	GAUDI2_SW_RST = PSOC_GLOBAL_CONF_RST_SRC_SW_RST_IND_SHIFT,
+	GAUDI2_FLR_RST = PSOC_GLOBAL_CONF_RST_SRC_FLR_RST_IND_SHIFT,
+	GAUDI2_ECC_DERR_RST = PSOC_GLOBAL_CONF_RST_SRC_ECC_DERR_RST_IND_SHIFT,
+};
+
+#endif /* __ASSEMBLY__ */
 #endif /* COMMON_GAUDI2_ARC_GNU_H_ */
diff --git a/gaudi2_arm/common.c b/gaudi2_arm/common.c
index 34044d3..5ee8ccf 100644
--- a/gaudi2_arm/common.c
+++ b/gaudi2_arm/common.c
@@ -543,3 +543,50 @@
 {
 	return mmPSOC_EFUSE_READ_CTRL;
 }
+
+u32 decode_reset_cause(void)
+{
+	u32 pos, rst_src, reg;
+
+	reg = regmap_readl(mmPSOC_GLOBAL_CONF_RST_SRC);
+
+	if (!reg)
+		return HL_UNKNOWN_RST;
+
+	pos = (__builtin_ffs(reg) - 1);
+
+	switch (pos) {
+	case GAUDI2_COLD_RST:
+		rst_src = HL_COLD_RST;
+		break;
+	case GAUDI2_PRSTN_RST:
+		rst_src = HL_PRSTN_RST;
+		break;
+	case GAUDI2_MANUAL_RST:
+		rst_src = HL_MANUAL_RST;
+		break;
+	case GAUDI2_WD_RST:
+		rst_src = HL_WD_RST;
+		break;
+	case GAUDI2_SW_RST:
+		rst_src = HL_SW_ALL_RST;
+		break;
+	case GAUDI2_SOFT_RST:
+		rst_src = HL_SOFT_RST;
+		break;
+	case GAUDI2_FW_RST:
+		rst_src = HL_FW_RST;
+		break;
+	case GAUDI2_FLR_RST:
+		rst_src = HL_FLR_RST;
+		break;
+	case GAUDI2_ECC_DERR_RST:
+		rst_src = HL_ECC_DERR_RST;
+		break;
+	default:
+		rst_src = HL_UNKNOWN_RST;
+		break;
+	}
+
+	return rst_src;
+}
diff --git a/gaudi2_arm/common.h b/gaudi2_arm/common.h
index 20fedfe..eea7081 100644
--- a/gaudi2_arm/common.h
+++ b/gaudi2_arm/common.h
@@ -7,6 +7,7 @@
 #define COMMON_GAUDI2_ARM_H_
 
 #include <gaudi2_blocks_linux_driver.h>
+#include <psoc_global_conf_masks.h>
 #include <bitops.h>
 
 #define HL_PCIE_LINK_RETRIES             50
@@ -91,5 +92,20 @@
 
 #define I2C_SLAVE_DEFAULT_INTR_MASK			0x48FF
 
+#ifndef __ASSEMBLY__
+
+enum gaudi2_rst_src {
+	GAUDI2_COLD_RST = PSOC_GLOBAL_CONF_RST_SRC_COLD_RST_IND_SHIFT,
+	GAUDI2_MANUAL_RST = PSOC_GLOBAL_CONF_RST_SRC_MNL_RST_IND_SHIFT,
+	GAUDI2_PRSTN_RST = PSOC_GLOBAL_CONF_RST_SRC_PRSTN_RST_IND_SHIFT,
+	GAUDI2_SOFT_RST = PSOC_GLOBAL_CONF_RST_SRC_SOFT_RST_IND_SHIFT,
+	GAUDI2_WD_RST = PSOC_GLOBAL_CONF_RST_SRC_WD_RST_IND_SHIFT,
+	GAUDI2_FW_RST = PSOC_GLOBAL_CONF_RST_SRC_FW_RST_IND_SHIFT,
+	GAUDI2_SW_RST = PSOC_GLOBAL_CONF_RST_SRC_SW_RST_IND_SHIFT,
+	GAUDI2_FLR_RST = PSOC_GLOBAL_CONF_RST_SRC_FLR_RST_IND_SHIFT,
+	GAUDI2_ECC_DERR_RST = PSOC_GLOBAL_CONF_RST_SRC_ECC_DERR_RST_IND_SHIFT,
+};
+
+#endif /* __ASSEMBLY__ */
 #endif /* COMMON_GAUDI2_ARM_H_ */
 
diff --git a/goya/common.c b/goya/common.c
index 08b0394..d69c54b 100644
--- a/goya/common.c
+++ b/goya/common.c
@@ -254,3 +254,33 @@
 	memcpy(hl_ts_data, goya_hl_ts_data,
 		sizeof(struct hl_ts_cfg) * GOYA_ASIC_TS_MAX_COUNT);
 }
+
+u32 decode_reset_cause(void)
+{
+	u32 rst_src;
+
+	rst_src = regmap_readl(mmPSOC_GLOBAL_CONF_RST_SRC);
+
+	switch (rst_src) {
+	case GOYA_COLD_RST:
+		rst_src = HL_COLD_RST;
+		break;
+	case GOYA_PRSTN_RST:
+		rst_src = HL_PRSTN_RST;
+		break;
+	case GOYA_MANUAL_RST:
+		rst_src = HL_MANUAL_RST;
+		break;
+	case GOYA_WD_RST:
+		rst_src = HL_WD_RST;
+		break;
+	case GOYA_SW_ALL_RST:
+		rst_src = HL_SW_ALL_RST;
+		break;
+	default:
+		rst_src = HL_UNKNOWN_RST;
+		break;
+	}
+
+	return rst_src;
+}
diff --git a/goya/common.h b/goya/common.h
index 1c489be..4086eb4 100644
--- a/goya/common.h
+++ b/goya/common.h
@@ -34,4 +34,15 @@
 #define I2C_SLAVE_ADDRESS_DBG	0x47
 #define I2C_SLAVE_DEFAULT_INTR_MASK	0x8FF
 
+#ifndef __ASSEMBLY__
+
+enum goya_rst_src {
+	GOYA_COLD_RST,
+	GOYA_MANUAL_RST,
+	GOYA_PRSTN_RST = 2,
+	GOYA_WD_RST = 4,
+	GOYA_SW_ALL_RST = 8,
+};
+
+#endif /* __ASSEMBLY__ */
 #endif /* COMMON_GOYA_H_ */
diff --git a/greco/common.c b/greco/common.c
index c0b2380..35a65c0 100644
--- a/greco/common.c
+++ b/greco/common.c
@@ -632,3 +632,33 @@
 	memcpy(hl_ts_data, greco_hl_ts_data,
 			sizeof(struct hl_ts_cfg) * HL_ASIC_TS_MAX_COUNT);
 }
+
+u32 decode_reset_cause(void)
+{
+	u32 rst_src;
+
+	rst_src = regmap_readl(mmPSOC_GLOBAL_CONF_RST_SRC);
+
+	switch (rst_src) {
+	case GRECO_COLD_RST:
+		rst_src = HL_COLD_RST;
+		break;
+	case GRECO_PRSTN_RST:
+		rst_src = HL_PRSTN_RST;
+		break;
+	case GRECO_MANUAL_RST:
+		rst_src = HL_MANUAL_RST;
+		break;
+	case GRECO_WD_RST:
+		rst_src = HL_WD_RST;
+		break;
+	case GRECO_SW_ALL_RST:
+		rst_src = HL_SW_ALL_RST;
+		break;
+	default:
+		rst_src = HL_UNKNOWN_RST;
+		break;
+	}
+
+	return rst_src;
+}
diff --git a/greco/common.h b/greco/common.h
index 5d0fd80..c9fb339 100644
--- a/greco/common.h
+++ b/greco/common.h
@@ -89,4 +89,16 @@
 #define PCIE_PHY_PROTOCOL0_EXT_RX_ADAPT_SEL_G4_CONF	0xFF
 
 #define PSOC_AXI_DRAIN_TIMEOUT				0x3E8 /* 1000 cycles */
+
+#ifndef __ASSEMBLY__
+
+enum greco_rst_src {
+	GRECO_COLD_RST,
+	GRECO_MANUAL_RST,
+	GRECO_PRSTN_RST = 2,
+	GRECO_WD_RST = 4,
+	GRECO_SW_ALL_RST = 8,
+};
+
+#endif /* __ASSEMBLY__ */
 #endif /* COMMON_GRECO_H_ */
diff --git a/greco_arc/common.c b/greco_arc/common.c
index 3889373..72075e3 100644
--- a/greco_arc/common.c
+++ b/greco_arc/common.c
@@ -363,3 +363,33 @@
 
 	return PREBOOT_TEXT_BASE;
 }
+
+u32 decode_reset_cause(void)
+{
+	u32 rst_src;
+
+	rst_src = regmap_readl(mmPSOC_GLOBAL_CONF_RST_SRC);
+
+	switch (rst_src) {
+	case GRECO_COLD_RST:
+		rst_src = HL_COLD_RST;
+		break;
+	case GRECO_PRSTN_RST:
+		rst_src = HL_PRSTN_RST;
+		break;
+	case GRECO_MANUAL_RST:
+		rst_src = HL_MANUAL_RST;
+		break;
+	case GRECO_WD_RST:
+		rst_src = HL_WD_RST;
+		break;
+	case GRECO_SW_ALL_RST:
+		rst_src = HL_SW_ALL_RST;
+		break;
+	default:
+		rst_src = HL_UNKNOWN_RST;
+		break;
+	}
+
+	return rst_src;
+}
diff --git a/greco_arc/common.h b/greco_arc/common.h
index 1eafa6e..88008b5 100644
--- a/greco_arc/common.h
+++ b/greco_arc/common.h
@@ -48,4 +48,15 @@
 #define ARC_RST_VALUE		0x110
 #define ARC_RST_VEC_ADDR	0x0
 
+#ifndef __ASSEMBLY__
+
+enum greco_rst_src {
+	GRECO_COLD_RST,
+	GRECO_MANUAL_RST,
+	GRECO_PRSTN_RST = 2,
+	GRECO_WD_RST = 4,
+	GRECO_SW_ALL_RST = 8,
+};
+
+#endif /* __ASSEMBLY__ */
 #endif /* COMMON_GRECO_ARC_H_ */
diff --git a/greco_arc_gnu/common.c b/greco_arc_gnu/common.c
index 0008810..58cfd16 100644
--- a/greco_arc_gnu/common.c
+++ b/greco_arc_gnu/common.c
@@ -243,3 +243,33 @@
 	memcpy(hl_ts_data, greco_hl_ts_data,
 			sizeof(struct hl_ts_cfg) * HL_ASIC_TS_MAX_COUNT);
 }
+
+u32 decode_reset_cause(void)
+{
+	u32 rst_src;
+
+	rst_src = regmap_readl(mmPSOC_GLOBAL_CONF_RST_SRC);
+
+	switch (rst_src) {
+	case GRECO_COLD_RST:
+		rst_src = HL_COLD_RST;
+		break;
+	case GRECO_PRSTN_RST:
+		rst_src = HL_PRSTN_RST;
+		break;
+	case GRECO_MANUAL_RST:
+		rst_src = HL_MANUAL_RST;
+		break;
+	case GRECO_WD_RST:
+		rst_src = HL_WD_RST;
+		break;
+	case GRECO_SW_ALL_RST:
+		rst_src = HL_SW_ALL_RST;
+		break;
+	default:
+		rst_src = HL_UNKNOWN_RST;
+		break;
+	}
+
+	return rst_src;
+}
diff --git a/greco_arc_gnu/common.h b/greco_arc_gnu/common.h
index f069c99..dba234c 100644
--- a/greco_arc_gnu/common.h
+++ b/greco_arc_gnu/common.h
@@ -43,4 +43,15 @@
 #define HL_SPI_BASE		0x1A000000
 #define HL_CFG_BASE		0xFC000000
 
+#ifndef __ASSEMBLY__
+
+enum greco_rst_src {
+	GRECO_COLD_RST,
+	GRECO_MANUAL_RST,
+	GRECO_PRSTN_RST = 2,
+	GRECO_WD_RST = 4,
+	GRECO_SW_ALL_RST = 8,
+};
+
+#endif /* __ASSEMBLY__ */
 #endif /* COMMON_GRECO_ARC_GNU_H_ */
diff --git a/include/preboot.h b/include/preboot.h
index 952a92a..7310531 100644
--- a/include/preboot.h
+++ b/include/preboot.h
@@ -50,9 +50,14 @@
 enum hl_rst_src {
 	HL_COLD_RST,
 	HL_MANUAL_RST,
-	HL_PRSTN_RST = 2,
-	HL_WD_RST = 4,
-	HL_SW_ALL_RST = 8,
+	HL_PRSTN_RST,
+	HL_WD_RST,
+	HL_SW_ALL_RST,
+	HL_SOFT_RST,
+	HL_FW_RST,
+	HL_FLR_RST,
+	HL_ECC_DERR_RST,
+	HL_UNKNOWN_RST = 0xFFF,
 };
 
 enum hl_debug_hw_func {
@@ -142,6 +147,7 @@
 void preboot_conf_iatu_access(bool enable);
 void preboot_disable_iatu_access(void);
 void hl_ram_init(void);
+u32 decode_reset_cause(void);
 
 #endif /* __ASSEMBLY__ */
 
diff --git a/ppboot/ppboot.c b/ppboot/ppboot.c
index 400aec5..9121cfd 100644
--- a/ppboot/ppboot.c
+++ b/ppboot/ppboot.c
@@ -106,7 +106,8 @@
 {
 	u32 rst_src;
 
-	rst_src = regmap_readl(mmPSOC_GLOBAL_CONF_RST_SRC);
+	rst_src = decode_reset_cause();
+
 	switch (rst_src) {
 	case HL_COLD_RST:
 	case HL_PRSTN_RST:
@@ -115,6 +116,7 @@
 	case HL_MANUAL_RST:
 	case HL_WD_RST:
 	case HL_SW_ALL_RST:
+	case HL_SOFT_RST:
 		/*
 		 * Currently, keep using the same flow as done before,
 		 * i.e. PCIe should already be initialized by default.
@@ -130,7 +132,8 @@
 
 static void rst_src_dump(void)
 {
-	unsigned int rst_src = regmap_readl(mmPSOC_GLOBAL_CONF_RST_SRC);
+	unsigned int rst_src = decode_reset_cause();
+
 	char *cause;
 
 	switch (rst_src) {
@@ -149,6 +152,19 @@
 	case HL_SW_ALL_RST:
 		cause = "software all";
 		break;
+	case HL_SOFT_RST:
+		cause = "soft";
+		break;
+	case HL_FW_RST:
+		cause = "firmware";
+		break;
+	case HL_FLR_RST:
+		cause = "function level";
+		break;
+	case HL_ECC_DERR_RST:
+		cause = "ecc derr";
+		break;
+	case HL_UNKNOWN_RST:
 	default:
 		cause = "unknown";
 		break;
